File: src/decode.js
// @ts-check

import { FALSE, TRUE, NULL, NUMBER, STRING, ARRAY, OBJECT, RECURSION, CUSTOM } from './constants.js';
import { I8, I16, I32, U8, U16, U32, LEN, BI, BUI } from './constants.js';

import { isArray, item, options, dv, v8 } from './utils.js';

const NUMBER_IGNORE = ~(RECURSION | NUMBER);
const CUSTOM_REVIVE = CUSTOM | I8;

const decoder = new TextDecoder;
const ignore = item(NULL, null);

/** @typedef {{ i: number }} Index */

/** @typedef {number[] | Uint8Array | import('./shared.js').default} Input */

/** @typedef {{ custom?: (value: unknown, encoded: boolean) => unknown }} Options */

/**
 * @param {Uint8Array} input
 * @param {Index} index
 * @param {boolean} isBigUint
 * @returns {bigint}
 */
const bigint = (input, index, isBigUint) => {
  for (let i = 0; i < 8; i++) v8[i] = input[index.i++];
  return isBigUint ? dv.getBigUint64(0, true) : dv.getBigInt64(0, true);
};

/**
 * @param {Uint8Array} input
 * @param {Index} index
 * @returns
 */
const floating = (input, index) => {
  for (let j = 4; j < 8; j++) v8[j] = input[index.i++];
  return dv.getFloat64(0, true);
};

/**
 * @param {Uint8Array} input
 * @param {Map<number, unknown>} cache
 * @param {Index} index
 * @returns
 */
const key = (input, cache, index) => {
  const type = input[index.i++];
  return (type & ~LEN) === RECURSION ?
    /** @type {string} */ (cache.get(number(input, type, index))) :
    string(input, cache, type, index);
};

/**
 * @param {Uint8Array} input
 * @param {number} type
 * @param {Index} index
 * @returns
 */
const number = (input, type, index) => {
  type &= NUMBER_IGNORE;

  if (type === 0) return type;

  v8[0] = input[index.i++];

  if (type === U8) return dv.getUint8(0);
  if (type === I8) return dv.getInt8(0);

  v8[1] = input[index.i++];
  if (type === U16) return dv.getUint16(0, true);
  if (type === I16) return dv.getInt16(0, true);

  v8[2] = input[index.i++];
  v8[3] = input[index.i++];
  if (type === U32) return dv.getUint32(0, true);
  if (type === I32) return dv.getInt32(0, true);

  return floating(input, index);
};

/**
 * @param {Uint8Array} input
 * @param {number} length
 * @param {Index} index
 * @returns
 */
const slice = (input, length, index) => {
  const i = index.i;
  index.i += length;
  return input.slice(i, i + length);
};

/**
 * @param {Uint8Array} input
 * @param {Map<number, unknown>} cache
 * @param {number} type
 * @param {Index} index
 * @returns {string}
 */
const string = (input, cache, type, index) => {
  if (type === STRING) return '';
  const known = index.i - 1;
  const length = number(input, type, index);
  const i = index.i;
  index.i += length;
  const str = decoder.decode(input.subarray(i, i + length));
  cache.set(known, str);
  return str;
};

/**
 * @param {Input} view
 * @param {Options} [options] 
 * @returns {unknown?}
 */
export const decode = (view, { custom = options.custom } = options) => {
  const input = isArray(view) ? new Uint8Array(view) : view;

  /** @type {Map<number, unknown>} */
  const cache = new Map;

  /** @type {Index} */
  const index = { i: 0 };

  /** @type {{ k: number, v: unknown }[]} */
  const stack = input.length ? [ignore] : [];

  let first = true, result, entry, prop;

  while (stack.length) {
    const { k, v } = stack.pop();

    if (k === OBJECT) prop = key(input, cache, index);

    const type = input[index.i++];

    if (type === FALSE) entry = false;
    else if (type === TRUE) entry = true;
    else if (type === NULL) entry = null;

    else if (CUSTOM <= type) {
      const length = number(input, input[index.i++], index);
      const view = slice(input, length, index);
      const revive = type === CUSTOM_REVIVE;
      entry = custom(revive ? decode(view) : view, !revive);
    }

    else if (type & NUMBER) {
      if (type & ARRAY) {
        const length = number(input, type & ~ARRAY, index);
        entry = slice(input, length, index);
      }
      else {
        const t = type & ~NUMBER;
        const isBigUint = t === BUI;
        if (isBigUint || t === BI) entry = bigint(input, index, isBigUint);
        else entry = number(input, type, index);
      }
    }

    else {
      const kind = type & ~LEN;
      if (kind === RECURSION) entry = cache.get(number(input, type, index));

      else if (kind === STRING) entry = string(input, cache, type, index);

      else {
        const known = index.i - 1;
        const length = number(input, type, index);

        if (kind === ARRAY) {
          entry = [];
          cache.set(known, entry);
          for (let next = item(ARRAY, entry), i = 0; i < length; i++)
            stack.push(next);
        }

        else if (kind === OBJECT) {
          entry = {};
          cache.set(known, entry);
          for (let next = item(OBJECT, entry), i = 0; i < length; i++)
            stack.push(next);
        }
      }
    }

    if (first) {
      first = false;
      result = entry;
    }
    else if (k === OBJECT) v[prop] = entry;
    else if (k === ARRAY) (/** @type {Array<unknown>} */ (v)).push(entry);
  }

  return result;
};

export default decode;

File: python/decode.py
"""
Decode flatted-view binary format (JS-compatible) to Python values.
"""
import struct

from .constants import (
    ARRAY,
    BI,
    BUI,
    CUSTOM,
    CUSTOM_REVIVE,
    F64,
    FALSE,
    I8,
    I16,
    I32,
    LEN,
    NULL,
    NUMBER,
    OBJECT,
    RECURSION,
    STRING,
    TRUE,
    U8,
    U16,
    U32,
)

NUMBER_IGNORE = ~(RECURSION | NUMBER)
# Same as JS `STACK_DECODED` in src/decode.js
STACK_DECODED = CUSTOM_REVIVE


def _default_custom(value, from_view=False):
    return value


def item(k, v):
    return {"k": k, "v": v}


def _read_number(data, type_byte, index):
    type_byte &= NUMBER_IGNORE
    if type_byte == 0:
        return 0
    b0 = data[index[0]]
    index[0] += 1
    if type_byte == U8:
        return b0
    if type_byte == I8:
        return struct.unpack("<b", bytes([b0]))[0]
    b1 = data[index[0]]
    index[0] += 1
    if type_byte == U16:
        return struct.unpack("<H", bytes([b0, b1]))[0]
    if type_byte == I16:
        return struct.unpack("<h", bytes([b0, b1]))[0]
    b2 = data[index[0]]
    b3 = data[index[0] + 1]
    index[0] += 2
    if type_byte == U32:
        return struct.unpack("<I", bytes([b0, b1, b2, b3]))[0]
    if type_byte == I32:
        return struct.unpack("<i", bytes([b0, b1, b2, b3]))[0]
    chunk = bytes([b0, b1, b2, b3]) + data[index[0] : index[0] + 4]
    index[0] += 4
    return struct.unpack("<d", chunk)[0]


def _read_bigint(data, index, is_unsigned):
    chunk = data[index[0] : index[0] + 8]
    index[0] += 8
    if is_unsigned:
        return struct.unpack("<Q", chunk)[0]
    return struct.unpack("<q", chunk)[0]


def _read_key(data, cache, index):
    type_byte = data[index[0]]
    index[0] += 1
    kind = type_byte & ~LEN
    if kind == RECURSION:
        idx = _read_number(data, type_byte, index)
        return cache[idx]
    return _read_string(data, cache, type_byte, index)


def _read_string(data, cache, type_byte, index):
    if type_byte == STRING:
        return ""
    known = index[0] - 1
    length = int(_read_number(data, type_byte, index))
    start = index[0]
    index[0] += length
    s = data[start : start + length].decode("utf-8")
    cache[known] = s
    return s


def _slice(data, length, index):
    start = index[0]
    index[0] += length
    return data[start : start + length]


def decode(view, custom=None):
    """Decode flatted-view binary to Python values.

    ``custom`` matches JS: ``(value, from_view: bool) -> unknown``.
    When ``from_view`` is True, ``value`` is the raw CUSTOM payload bytes from ``view(...)``.
    When False, ``value`` is the list built by decoding the implicit CUSTOM array segment.
    """
    if custom is None:
        custom = _default_custom

    data = bytes(view)

    cache = {}
    # CUSTOM placeholders: list object -> byte offset (JS Map entry-as-key)
    rev = {}
    as_key = set()

    index = [0]
    ignore = item(NULL, None)
    stack = [ignore] if len(data) > 0 else []
    first = True
    result = None
    entry = None
    prop = None

    def finalize(k, pair, known):
        nonlocal first, result, entry
        value, inner = pair[0], pair[1]
        key = inner["k"]
        parent = inner["v"]
        revived = custom(value, k != STACK_DECODED)
        cache[known] = revived
        if parent is not None:
            parent[key] = revived
        elif first:
            result = revived
        return revived

    while stack:
        frame = stack.pop()
        k, v = frame["k"], frame["v"]

        if k == OBJECT:
            prop = _read_key(data, cache, index)

        elif CUSTOM <= k and isinstance(v, list):
            custom_pair = v
            known_ref = rev.get(id(custom_pair[0]))
            custom_pair[0].append(finalize(k, custom_pair, known_ref))
            continue

        if index[0] >= len(data):
            break

        type_byte = data[index[0]]
        index[0] += 1

        if type_byte == FALSE:
            entry = False
        elif type_byte == TRUE:
            entry = True
        elif type_byte == NULL:
            entry = None

        elif CUSTOM <= type_byte:
            known = index[0] - 1
            entry = []
            cache[known] = entry
            rev[id(entry)] = known
            as_key.add(id(entry))
            inner_k = None
            if k == OBJECT:
                inner_k = prop
            elif k == ARRAY:
                inner_k = len(v)
            stack.append(
                item(
                    type_byte,
                    item(inner_k, v),
                )
            )
            continue

        elif type_byte & NUMBER:
            if type_byte & ARRAY:
                length = int(_read_number(data, type_byte & ~ARRAY, index))
                entry = _slice(data, length, index)
            else:
                t = type_byte & ~NUMBER
                if t == BUI:
                    entry = _read_bigint(data, index, True)
                elif t == BI:
                    entry = _read_bigint(data, index, False)
                else:
                    entry = _read_number(data, type_byte, index)

        else:
            kind = type_byte & ~LEN
            if kind == RECURSION:
                idx = _read_number(data, type_byte, index)
                entry = cache[idx]
            elif kind == STRING:
                entry = _read_string(data, cache, type_byte, index)
            else:
                known = index[0] - 1
                length = int(_read_number(data, type_byte, index))
                is_custom = CUSTOM <= k

                if kind == ARRAY:
                    if is_custom:
                        entry = cache.get(known - 1)
                        if k != STACK_DECODED:
                            target = entry
                            pair = [_slice(data, length, index), v]
                            target.append(finalize(k, pair, known - 1))
                            continue
                        stack.append(item(k, [entry, v]))
                    else:
                        entry = []
                        cache[known] = entry
                    for _ in range(length):
                        stack.append(item(ARRAY, entry))

                elif kind == OBJECT:
                    entry = {}
                    cache[known] = entry
                    for _ in range(length):
                        stack.append(item(OBJECT, entry))

        if first:
            first = False
            result = entry
        elif k == OBJECT:
            v[prop] = entry
        elif k == ARRAY:
            v.append(entry)

    if isinstance(result, list) and id(result) in as_key:
        return result.pop()
    return result
